### ISSUE TITLE
reheader --in-place tests fail on MIPS and ARM little endian architectures

### ISSUE THREAD

**Issue by satta** [samtools/samtools#526]
When building **samtools** 1.3 on Debian for the mipsel and armel architectures, these two tests fail (as the only ones):

```
test_reheader:
	/«PKGBUILDDIR»/samtools reheader --in-place /«PKGBUILDDIR»/test/reheader/hdr.sam /«PKGBUILDDIR»/test/dat/view.001.tmp.v21.cram && /«PKGBUILDDIR»/samtools view -h /«PKGBUILDDIR»/test/dat/view.001.tmp.v21.cram | perl -pe 's/	VN:.*//'

	ERR: 
.. failed ...

test_reheader:
	/«PKGBUILDDIR»/samtools reheader --in-place /«PKGBUILDDIR»/test/reheader/hdr.sam /«PKGBUILDDIR»/test/dat/view.001.tmp.v30.cram && /«PKGBUILDDIR»/samtools view -h /«PKGBUILDDIR»/test/dat/view.001.tmp.v30.cram | perl -pe 's/	VN:.*//'

	ERR: 
.. failed ...
```

Any idea what might be the cause?

**Comment by domibel** [comment]
Not sure, but I suggest to fix all htslib errors first. This one here might be a side effect of something else.

**Comment by satta** [comment]
Thanks for looking into it. Please note that this build failure is currently preventing samtools to migrate to Debian testing. If you consider this a bug, the build logs are [here](https://buildd.debian.org/status/package.php?p=samtools). Otherwise, if you consider these architectures unsupported or not a high priority, I can remove these tests from the affected architectures for now.

**Comment by daviesrob** [comment]
It looks like this is due to **samtools** being built with `-D_FILE_OFFSET_BITS=64 -D_LARGEFILE64_SOURCE` while **htslib** was built without them.  This means the size of `off_t` is different between the two, making a call to `hseek()` pass over the wrong parameters.

The solution is either to ensure the above two macros are set when compiling **htslib**, or possibly better run htslib's optional configure script before building it, which does ensure that they are set.

**Comment by jmarshall** [comment]
Since 1.3, I've added a configure test in htslib to add `-D_FILE_OFFSET_BITS=64` where appropriate: samtools/htslib@f1327a703583bf61e2e7ffd11198487fee2a8422.  You should be able to pull this one commit in **and run `configure`** in htslib and have these tests work.

Sadly big-endian platforms are a low priority for us as most users are on the dominant x86 platforms.  We would like samtools to work elsewhere too, but at the moment I think there is still a reasonable amount of code that assumes little-endian and so samtools doesn't fully work on big-endian platforms.

Soon I will be adding a similar configure test in samtools, so samtools will no longer always have `‑D_FILE_OFFSET_BITS=64 -D_LARGEFILE64_SOURCE` in its makefile.  Then these will mostly always be set or unset in parallel; and 32-bit platforms will **need** to run configure to access large files.

**Comment by satta** [comment]
Thanks. I have added the patch and will make sure our samtools runs its tests with an updated htslib. Will let you know how it goes.

**Comment by satta** [comment]
Excellent, tests are succeeding now.
